### ISSUE TITLE
Parser should support PostgreSQL date part syntaxes with explicit type casts

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12645]
When the parser expects a literal (string, number, etc), we currently don't allow explicit casts, but we could support them. One use-case is when PostgreSQL users use function based indexes:

```sql
create table t (ts timestamp);
create index i on t(date_trunc('day', ts));
```

The SQL expression is transformed to this:

```sql
date_trunc('day'::text, ts)
```

As can be seen with this query:

```sql
select pg_get_indexdef(i.indexrelid, 1, true)
from pg_index i
join pg_class t on t.oid = i.indrelid
where t.relname = 't'
```

But the parser doesn't support this syntax, currently:

```
$ jbang parser@jooq
> select date_trunc('day'::text, ts);
Token ',' expected: [1:24] select date_trunc('day'[*]::text, ts);
```

**Comment by lukaseder** [comment]
The date part parsing logic is quite hard-wired, so supporting various cast syntaxes might be harder. Let's just make this about the PG specific syntax.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16 and 3.15.5 (https://github.com/jOOQ/jOOQ/issues/12646). A fix in 3.14.16 was planned (https://github.com/jOOQ/jOOQ/issues/12647) but retracted, as 3.14 seems to be unable to parse date parts in string literals.
